Application/targets/STM32/Core/Src/main.c

```diff
@@ -373,26 +451,66 @@ static void MX_GPIO_Init(void)
   HAL_NVIC_SetPriority(EXTI0_IRQn, 0, 0);
   HAL_NVIC_EnableIRQ(EXTI0_IRQn);
 
+/* USER CODE BEGIN MX_GPIO_Init_2 */
+/* USER CODE END MX_GPIO_Init_2 */
 }
 
 /* USER CODE BEGIN 4 */
 
 // EXTI Line0 External Interrupt ISR Handler CallBackFun
+// Starts interrupt timer when a zero crossing event occurs.
 void HAL_GPIO_EXTI_Callback(uint16_t GPIO_Pin)
 {
     if(GPIO_Pin == ZC_IN_Pin) // If The INT Source Is EXTI Line0 (A0 Pin)
     {
-    	num_zero_crossings++;
-    	HAL_GPIO_WritePin(GPIOB, TRIAC_FIRE_Pin, GPIO_PIN_SET); // Reset triac
-    	delay_us(triac_fire_delay_us, &htim1);
+    	HAL_TIM_Base_Start_IT(&htim2);
+    	num_zero_crossings++; // For debug
+    	//delay_us(triac_fire_delay_us, &htim1);
+    	//HAL_GPIO_WritePin(GPIOB, TRIAC_FIRE_Pin, GPIO_PIN_RESET); // Fire triac
+    }
+}
+
+// Fires triac when timer 2 overflows then disables timer 2 until next zero crossing.
+// timer 2 overflow must always occur in less than 8.333ms after timer 2 starts.
+
+#define TRIAC_EN_DURATION_US 50;
+
+void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim)
+{
+	uint16_t timer2_val = __HAL_TIM_GET_COUNTER(&htim2); // for debug
+	// Timer overflow interrupt callback
+    if (htim->Instance == TIM2)
+    {
+    	timer2_val = __HAL_TIM_GET_COUNTER(&htim2); // for debug
+    	HAL_TIM_Base_Stop(&htim2); // Stop timer 2
+    	//HAL_TIM_Base_Stop_IT(&htim2); // Stop timer 2 IT
+    	TIM2->CNT = 0; // Reset timer 2 to zero
+    	timer2_val = __HAL_TIM_GET_COUNTER(&htim2); // for debug
+
+    	//HAL_TIM_Base_Stop_IT(&htim2);
+    	//HAL_TIM_Base_Stop(&htim2); // Stop timer 2. This function also resets the timer to zero.
+    	//timer2_val = __HAL_TIM_GET_COUNTER(&htim2); // for debug
```

Comment:
Is there good reason to let this commented code hang out? If it isn't going to be used frequently I would lean toward removing it.